src/target/lpc55xx.c

```diff
@@ -37,12 +37,478 @@
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_INIT (%d)\n", status);
+		goto exit;
+	}
+
+	target_mem_read(t, config, LPC55XX_FLASH_CONFIG_ADDRESS, sizeof(*config));
+
+	success = true;
+
+exit:
+
+	target_mem_write(t, LPC55XX_FLASH_CONFIG_ADDRESS, backup_memory, sizeof(backup_memory));
+	target_regs_write(t, regs);
+
+	return success;
+}
+
+static bool lpc55xx_get_uuid(target *t, uint8_t *uuid)
+{
+	uint8_t backup_memory[LPC55XX_SCRATCH_MEMORY_LEN + LPC55XX_UUID_LEN];
+	uint32_t regs[t->regs_size / sizeof(uint32_t)];
+
+	target_regs_read(t, regs);
+	target_mem_read(t, backup_memory, LPC55XX_FLASH_CONFIG_ADDRESS, sizeof(backup_memory));
+
+	bool success = false;
+
+	lpc55xx_prepare_flash_config(t, LPC55XX_FLASH_CONFIG_ADDRESS);
+
+	enum iap_status status;
+
+	status = iap_call_raw(t, IAP_CMD_FLASH_INIT, 0, 0, 0);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_INIT (%d)\n", status);
+		goto exit;
+	}
+
+	status = iap_call_raw(t, IAP_CMD_FFR_INIT, 0, 0, 0);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FFR_INIT (%d)\n", status);
+		goto exit;
+	}
+
+	status = iap_call_raw(t, IAP_CMD_FFR_GET_UUID, LPC55XX_UUID_ADDRESS, 0, 0);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FFR_GET_UUID (%d)\n", status);
+		goto exit;
+	}
+
+	target_mem_read(t, uuid, LPC55XX_UUID_ADDRESS, LPC55XX_UUID_LEN);
+
+	success = true;
+
+exit:
+
+	target_mem_write(t, LPC55XX_FLASH_CONFIG_ADDRESS, backup_memory, sizeof(backup_memory));
+	target_regs_write(t, regs);
+
+	return success;
+}
+
+static bool lpc55xx_flash_erase(target_flash_s *tf, target_addr_t addr, size_t len)
+{
+	enum iap_status status = iap_call_raw(tf->t, IAP_CMD_FLASH_ERASE, addr, (uint32_t)len, LPC55XX_ERASE_KEY);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_ERASE (%d)\n", status);
+		return false;
+	}
+
+	return true;
+}
+
+static bool lpc55xx_flash_write(target_flash_s *tf, target_addr_t dest, const void *src, size_t len)
+{
+	target_mem_write(tf->t, LPC55XX_WRITE_BUFFER_ADDRESS, src, len);
+
+	enum iap_status	status = iap_call_raw(tf->t, IAP_CMD_FLASH_PROGRAM, dest, LPC55XX_WRITE_BUFFER_ADDRESS, (uint32_t)len);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_PROGRAM (%d)\n", status);
+		return false;
+	}
+
+	return true;
+}
+
+static bool lpc55xx_enter_flash_mode(target *t)
+{
+	// TODO: do we need to reset the target here?
+	// (reset here doesn't work for some reason)
+	// target_reset(t);
+
+	lpc55xx_prepare_flash_config(t, LPC55XX_FLASH_CONFIG_ADDRESS);
+
+	enum iap_status status;
+
+	// Initialize the IAP flash context once in a predefined location
+	// of SRAM, the flash erase/write functions assume it is present.
+
+	status = iap_call_raw(t, IAP_CMD_FLASH_INIT, 0, 0, 0);
+
+	if (status != IAP_STATUS_FLASH_SUCCESS) {
+		DEBUG_WARN("LPC55xx: IAP error: FLASH_INIT (%d)\n", status);
+		return false;
+	}
```

Comment:
In some ways it kind of fits in both, because it initializes the flash controller AND writes a structure in RAM. But I agree I think `prepare` would be the better place for it, I will try to move it.